File: 465.optimal-account-balancing.cpp
/*
 * @lc app=leetcode id=465 lang=cpp
 *
 * [465] Optimal Account Balancing
 */
#include <bits/stdc++.h>

using std::unordered_map;
using std::vector;

// @lc code=start
// "settle the debt" means no one owes others

class Solution {
public:
    // backtracking
    // T: O((n - 1)!), n := len(transactions)
    // S: O(n)
    int minTransfers(vector<vector<int>>& transactions) {
        unordered_map<int, int> balances;
        for (const auto& tran : transactions) {
            balances[tran[0]] -= tran[2];
            balances[tran[1]] += tran[2];
        }
        vector<int> debts;
        for (const auto& p : balances) {
            if (p.second != 0) {
                // debts[i] > 0 means someone owes others;
                // debts[i] < 0 means someone needs money returned by others
                debts.push_back(p.second);
            }
        }

        return dfs(0, debts);
    }

private:
    int dfs(int start_idx, vector<int>& debts) {
        const int n = debts.size();

        // skip the zero debts
        while (start_idx < n && debts[start_idx] == 0) {
            ++start_idx;
        }

        if (start_idx == n) {
            return 0;
        }

        int ans = INT_MAX;
        for (int i = start_idx + 1; i < n; ++i) {
            if (debts[i] * debts[start_idx] < 0) {   // different signs, one owes other
                debts[i] += debts[start_idx];
                ans = std::min(ans, 1 + dfs(start_idx + 1, debts));
                debts[i] -= debts[start_idx];  // backtrack
            }
        }
        return ans;
    }
};
// @lc code=end

File: 465.optimal-account-balancing.py
#
# @lc app=leetcode id=465 lang=python3
#
# [465] Optimal Account Balancing
#
from collections import defaultdict
from typing import List

# @lc code=start
class Solution:
    # Memoized DFS + DP + bit manipulation
    # T: O(n * 2 ^ n)
    # S: O(2 ^ n)
    def minTransfers(self, transactions: List[List[int]]) -> int:
        balances = defaultdict(int)
        for u, v, amount in transactions:
            balances[u] += amount
            balances[v] -= amount

        debts = [x for x in balances.values() if x != 0]
        n = len(debts)

        """
        Find m, the max number of disjoint sets where each set s has sum(s) == 0.
        Each of these sets CANNOT be divided to setA and setB where sum(setA) == sum(setB) == 0.
        For such each set we need len(s) - 1 transactions to settle the debt.
            ans = len(s_1) - 1 + len(s_2) - 1 + ... + len(s_m) - 1
            ans = len(s_1) + len(s_2) + ... + len(s_m) - 1 * m
        The final answer is n - m.
        """

        # k: bitmask representing what people are involved when building the sets, 
        # v: the corresponding max number of disjoint sets where each set has zero balance sum. 
        memo = [-1] * (1 << n)  
        memo[0] = 0

        def dfs(mask):
            if memo[mask] != -1:
                return memo[mask]
            
            balance_sum = 0
            ans = 0

            # remove one person at a time
            for i in range(n):
                person_mask = 1 << i
                if mask & person_mask:
                    balance_sum += debts[i]
                    ans = max(ans, dfs(mask ^ person_mask))  # use XOR to remove the person

            memo[mask] = ans + (balance_sum == 0)
            return memo[mask]
        
        return n - dfs((1 << n) - 1)

# @lc code=end

